### ISSUE TITLE
Optimize rewrite rules

### ISSUE THREAD

**Issue by olafurpg** [olafurpg/scalafmt#465]
It seems like the new [rewrite.rules](https://olafurpg.github.io/scalafmt/#rewrite.rules) add a ~30-70% performance penalty formatting individual files, depending on the file size

```

[info] Micro.ExtraLarge.scalafmt          avgt   10  2782.812 ± 173.291  ms/op
[info] Micro.ExtraLarge.scalafmt_rewrite  avgt   10  4979.544 ± 302.429  ms/op
[info] Micro.Medium.scalafmt              avgt   10    95.079 ±   6.821  ms/op
[info] Micro.Medium.scalafmt_rewrite      avgt   10   115.916 ±   3.811  ms/op
[info] Micro.Small.scalafmt               avgt   10    14.713 ±   3.636  ms/op
[info] Micro.Small.scalafmt_rewrite       avgt   10    10.018 ±   2.554  ms/op

```

To run the benchmark, clone the repo and

```
/run-benchmarks.sh Micro 10
```

I suspect micro optimization can shrink that gap to <5% overhead. For example, the implementation of `Patch` is super naive: https://github.com/olafurpg/scalafmt/blob/50affca0024e1b98cbaa75bc4de662f4140b1d44/core/src/main/scala/org/scalafmt/rewrite/Patch.scala#L15-L36
It builds a new array of ALL tokens (including every space/newline) in the file for every single patch. I think it is possible to apply all of the patches in a single run.

**Comment by JesusMtnez** [comment]
We are going to try and look at Lambda World 2016 Hackathon.

**Comment by olafurpg** [comment]
This was fixed in #706
